File: MatchIQ_Flutter/lib/data/api/auth_api.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../../core/constants/app_constants.dart';

class AuthToken {
  AuthToken({
    required this.accessToken,
    required this.userId,
    required this.userUuid,
    required this.displayName,
    this.email,
  });

  final String accessToken;
  final int userId;
  final String userUuid;
  final String displayName;
  final String? email;

  factory AuthToken.fromJson(Map<String, dynamic> json, {String? email}) {
    return AuthToken(
      accessToken: json['access_token'] as String,
      userId: json['user_id'] as int,
      userUuid: json['user_uuid'] as String,
      displayName: (json['display_name'] as String?) ?? 'Player',
      email: email,
    );
  }
}

/// Real Game DB auth via Backend FastAPI `/api/v1/auth/*`
class AuthApi {
  AuthApi({http.Client? client}) : _client = client ?? http.Client();

  final http.Client _client;

  Uri _uri(String path) => Uri.parse('${AppConstants.apiBaseUrl}$path');

  Future<AuthToken> register({
    required String email,
    required String password,
    required String displayName,
  }) async {
    final res = await _client.post(
      _uri('/auth/register'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'email': email,
        'password': password,
        'display_name': displayName,
      }),
    );
    return _parse(res, email: email, fallback: 'Register failed');
  }

  Future<AuthToken> login({
    required String email,
    required String password,
  }) async {
    final res = await _client.post(
      _uri('/auth/login'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'email': email, 'password': password}),
    );
    return _parse(res, email: email, fallback: 'Login failed');
  }

  Future<AuthToken> guest({String displayName = 'Guest'}) async {
    final guestId =
        'guest-${DateTime.now().millisecondsSinceEpoch}-${DateTime.now().microsecond}';
    final res = await _client.post(
      _uri('/auth/guest'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'guest_id': guestId,
        'display_name': displayName,
      }),
    );
    return _parse(res, fallback: 'Guest login failed');
  }

  Future<AuthToken> google({
    required String googleId,
    required String email,
    required String displayName,
  }) async {
    final res = await _client.post(
      _uri('/auth/google'),
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({
        'google_id': googleId,
        'email': email,
        'display_name': displayName,
      }),
    );
    return _parse(res, email: email, fallback: 'Google login failed');
  }

  AuthToken _parse(
    http.Response res, {
    String? email,
    required String fallback,
  }) {
    final body = res.body.isEmpty ? <String, dynamic>{} : jsonDecode(res.body);
    if (res.statusCode >= 200 && res.statusCode < 300) {
      return AuthToken.fromJson(Map<String, dynamic>.from(body as Map), email: email);
    }
    final detail = body is Map ? body['detail'] : null;
    if (detail is String) throw Exception(detail);
    if (detail is List && detail.isNotEmpty) {
      final first = detail.first;
      if (first is Map && first['msg'] != null) {
        throw Exception(first['msg'].toString());
      }
    }
    throw Exception(fallback);
  }
}

File: MatchIQ_App/src/api/authApi.ts
import { apiClient } from './client';
import type { AuthSession, PlayerProfile } from '../types';

export type TokenResponse = {
  access_token: string;
  token_type: string;
  user_id: number;
  user_uuid: string;
  display_name: string;
};

function toSession(token: TokenResponse, email?: string): AuthSession {
  const user: PlayerProfile = {
    id: String(token.user_id),
    username: email?.split('@')[0] || token.display_name.toLowerCase().replace(/\s+/g, '_'),
    displayName: token.display_name || 'Player',
    avatarId: 'avatar_default',
    frameId: 'frame_gold',
    level: 1,
    xp: 0,
    xpToNext: 1000,
    rank: 'Rookie',
    referralCode: token.user_uuid.slice(0, 8).toUpperCase(),
    wins: 0,
    losses: 0,
    winRate: 0,
  };
  return {
    token: token.access_token,
    refreshToken: token.access_token,
    user,
  };
}

function apiError(error: unknown, fallback: string): Error {
  const ax = error as { response?: { data?: { detail?: string | { msg?: string }[] } }; message?: string };
  const detail = ax.response?.data?.detail;
  if (typeof detail === 'string') return new Error(detail);
  if (Array.isArray(detail) && detail[0]?.msg) return new Error(detail[0].msg);
  return new Error(ax.message || fallback);
}

/** Real Game DB auth via Backend FastAPI `/api/v1/auth/*` */
export const authApi = {
  async register(email: string, password: string, displayName: string): Promise<AuthSession> {
    try {
      const { data } = await apiClient.post<TokenResponse>('/auth/register', {
        email,
        password,
        display_name: displayName || 'Player',
      });
      return toSession(data, email);
    } catch (e) {
      throw apiError(e, 'Register failed');
    }
  },

  async login(email: string, password: string): Promise<AuthSession> {
    try {
      const { data } = await apiClient.post<TokenResponse>('/auth/login', { email, password });
      return toSession(data, email);
    } catch (e) {
      throw apiError(e, 'Login failed');
    }
  },

  async guest(displayName = 'Guest'): Promise<AuthSession> {
    try {
      const guestId = `guest-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
      const { data } = await apiClient.post<TokenResponse>('/auth/guest', {
        guest_id: guestId,
        display_name: displayName,
      });
      return toSession(data);
    } catch (e) {
      throw apiError(e, 'Guest login failed');
    }
  },

  async google(googleId: string, email: string, displayName: string): Promise<AuthSession> {
    try {
      const { data } = await apiClient.post<TokenResponse>('/auth/google', {
        google_id: googleId,
        email,
        display_name: displayName,
      });
      return toSession(data, email);
    } catch (e) {
      throw apiError(e, 'Google login failed');
    }
  },

  async me(): Promise<{
    id: number;
    user_uuid: string;
    email: string | null;
    display_name: string;
    is_guest: boolean;
    avatar_url: string | null;
  }> {
    const { data } = await apiClient.get('/auth/me');
    return data;
  },
};
